netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -12,6 +12,68 @@
 #include "ebpf_store_helper.h"
 #include "net_ebpf_ext_sock_addr.h"
 
+#define TARGET_PROCESS_ID 1234
+#define EXPIRY_TIME 60000 // 60 seconds in ms.
+#define CONVERT_100NS_UNITS_TO_MS(x) ((x) / 10000)
+
+/**
+ * Connection context info does not contain the source IP address because
+ * the source IP address is not always available at connect_redirect layer.
+ * Source port is however available and included below for a stricter check.
+ */
+typedef struct _net_ebpf_ext_connect_context_address_info
+{
+    uint32_t family;
+    union
+    {
+        uint32_t ipv4;
+        uint32_t ipv6[4];
+    } destination_ip;
+    uint16_t destination_port;
+    uint16_t source_port;
+} net_ebpf_ext_connect_context_address_info_t;
+
+typedef struct _net_ebpf_ext_connection_context_key
+{
+    uint64_t transport_endpoint_handle;
+    net_ebpf_ext_connect_context_address_info_t address_info;
+    uint32_t compartment_id;
+    uint16_t protocol;
+} net_ebpf_ext_connection_context_key_t;
+
+typedef struct _net_ebpf_extension_connection_context
+{
+    LIST_ENTRY list_entry;
+    net_ebpf_ext_connection_context_key_t key;
+    struct
+    {
+        uint8_t verdict : 2;
+        uint8_t redirected : 1;
+    } value;
+    uint64_t timestamp;
+} net_ebpf_extension_connection_context_t;
+
+typedef struct _net_ebpf_extension_redirect_handle_entry
+{
+    LIST_ENTRY list_entry;
+    uint64_t filter_id;
+    HANDLE redirect_handle;
+} net_ebpf_extension_redirect_handle_entry_t;
+
+typedef struct _net_ebpf_ext_sock_addr_statistics
+{
+    volatile long permit_connection_count;
+    volatile long redirect_connection_count;
+    volatile long block_connection_count;
+} net_ebpf_ext_sock_addr_statistics_t;
+
+static net_ebpf_ext_sock_addr_statistics_t _net_ebpf_ext_statistics;
+
+static EX_SPIN_LOCK _net_ebpf_ext_sock_addr_lock;
+_Guarded_by_(_net_ebpf_ext_sock_addr_lock) static LIST_ENTRY _net_ebpf_ext_redirect_handle_list;
+_Guarded_by_(_net_ebpf_ext_sock_addr_lock) static LIST_ENTRY _net_ebpf_ext_connect_context_list;
```

Comment:
If you intend to move this to a hashtable in future, please create an issue and refer it here.